### PR TITLE
Support for passing Crystal objects to Lua and call methods on them

### PR DESCRIPTION
**Pull request by jan-zajic**
Implemented using Lua userdata and metatables. This also resolves issue #14. See callable_support_spec.cr and examples/callable.cr for demonstration of all features.

I will be pleased to discuss this large work.

### CODE REVIEW

**Comment by veelenga** [comment]
Looks promising. I will need some time to understand the implementation.

File: examples/callable.cr [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+      @animals.each do |a|
+        if a.hungry > 0
+          io << a.name << ": " << a.hungry << "\n"
+        end
+      end
+    end
+  end
+
+  def feed_all
+    @animals.each do |a|
```
Comment (veelenga) [inline examples/callable.cr:89]: `@animals.each &.feed`

File: src/lua/callable.cr [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+          {% end %}
+        {% end %}
+      {% end %}
+      {% for k, m in method_map %}
+          {% if !m.name.ends_with?("=") && !m.name.starts_with?("_") %}
+          if key == "{{m.name}}"
+              proc = ->(state : LibLua::State) {
+                stack = Lua::Stack.new(state, :all)
+                {% reverse_args = [] of Arg %}
+                {% for a, index in m.args %}
```
Comment (veelenga) [inline src/lua/callable.cr:90]:
```suggestion
                {% for a in m.args %}
```

File: src/lua/callable.cr [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                  {% reverse_args = [a] + reverse_args %}
```
Comment (veelenga) [inline src/lua/callable.cr:91]:
```suggestion
                  {% reverse_args = reverse_args.unshift(a) %}
```

File: src/lua/callable.cr [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+                {% end %}
+                {% for a, index in reverse_args %}
+                    {% if a.restriction.is_a?(Nop) %}
+                        {{a.name}} = stack[-{{index + 1}}]
```
Comment (veelenga) [inline src/lua/callable.cr:95]: Could you please elaborate. What do we want to do if argument is_a? Nop here?

File: src/lua/callable.cr [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+      super
+      {% end %}
+      {% end %}
+    end
+  end
+
+  macro included
+    _lua_new_instance
+
+    macro finished
```
Comment (veelenga) [inline src/lua/callable.cr:121]: Nice

File: src/lua/object.cr [hunk 6]
```diff
@@ -38,4 +38,15 @@ module Lua
       end
     end
   end
+
+  class Reference
```
Comment (veelenga) [inline src/lua/object.cr:42]: Why can't we re-use `Lua::Object`? It acts very similar to the reference.

File: src/lua/stack.cr [hunk 7]
```diff
@@ -76,18 +77,26 @@ module Lua
...
+      when Symbol                     then LibLua.pushstring(@state, o.to_s)
+      when Array, Tuple               then pushtable(o.to_a)
+      when Hash, NamedTuple           then pushtable(o.to_h)
+      when Proc(LibLua::State, Int32) then pushclosure(o.as(Proc))
+      when Class
+        if o < LuaCallable
+          pushmetatable(o)
+        else
+          raise ArgumentError.new(
+            "unable to pass Crystal Class of type '#{typeof(o)}' to Lua"
```
Comment (veelenga) [inline src/lua/stack.cr:95]: please add a test for this.

File: src/lua/stack/type.cr [hunk 8]
```diff
@@ -34,5 +34,41 @@ module Lua
     def typename(type : TYPE)
       String.new LibLua.typename(@state, type.value)
     end
+
+    def crystal_type_at(pos : Int)
```
Comment (veelenga) [inline src/lua/stack/type.cr:38]: Would you mind to add some method docs with examples? This is a public API and would be nice to have some hints

File: spec/lua/stack/callable_support_spec.cr [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        s.set_global("cs", CallableSubClass)
+        s.set_global("csi", CallableSubClass.new)
+        s.run %q{
+          csi2 = cs.new()
+          assert(csi.subclass_function() == "Hi i'm CallableSubClass. This is my own function.", "call subclass method wrong res: " .. csi.subclass_function())
+          assert(csi.simple_function() == "Hello from CallableSubClass", "call overriden method wrong res: " .. csi.simple_function())
+          assert(csi2.simple_function() == "Hello from CallableSubClass", "call overriden method wrong res: " .. csi2.simple_function())
+          csi.w = "First"
+          csi2.w = "Second"
+          assert(csi.join_other(csi2) == "Second after First", "call inherited method wrong res: " .. csi.join_other(csi2))
```
Comment (veelenga) [inline spec/lua/stack/callable_support_spec.cr:74]: assertion within Lua to verify test results. Nice idea 👍

**Review by veelenga — COMMENTED** [review]
Cool. The macro part is kinda hard to read and understand, but overall this is just a great job.  

Sorry for the late response.

**Comment by veelenga** [comment]
Going to merge this and address comments on my own. Thanks

**Comment by jan-zajic** [comment]
@veelenga I am very sorry for not responding, my free time is fully filled with work on Crystal language itself for now. I kept postponing this PR all the time.. But I will return to Lua integration as soon as Crystal compiler work on Windows in official master branch (currently only on my PC;). Thank you very much for merging this and make it clean.

**Pull request merged**